crates/burn-cubecl/src/kernel/index/slice_assign.rs

```diff
@@ -1,59 +1,112 @@
 use crate::{CubeRuntime, element::CubeElement, tensor::CubeTensor};
 use cubecl::{calculate_cube_count_elemwise, prelude::*};
+use cubecl_std::{FastDivmod, FastDivmodArgs};
 use std::ops::Range;
 
 #[cube(launch)]
 fn slice_assign_kernel<E: CubePrimitive>(
-    input: &mut Tensor<E>,
-    value: &Tensor<E>,
-    indices: Sequence<u32>,
-    #[comptime] rank: u32,
+    input: &mut Tensor<Line<E>>,
+    value: &Tensor<Line<E>>,
+    slice_shape: Sequence<FastDivmod>,
+    slice_offsets: Sequence<u32>,
 ) {
+    if ABSOLUTE_POS >= value.len() {
+        terminate!()
+    }
+
+    let rank = comptime!(slice_shape.len());
+
+    let line_size = input.line_size();
+    let mut offset_remainder = ABSOLUTE_POS * line_size;
     let mut offset_input = 0;
     let mut offset_value = 0;
 
+    let mut i = comptime![0];
+
+    #[allow(clippy::explicit_counter_loop)]
     #[unroll]
-    for i in 0..rank {
-        let range_start = *indices.index(i);
-        let offset_local = ABSOLUTE_POS / value.stride(i) % value.shape(i);
+    for _ in 0..rank {
+        let dim = comptime![rank - i - 1];
+        let (rem, offset_local) = slice_shape.index(dim).div_mod(offset_remainder);
+
+        let range_start = *slice_offsets.index(dim);
         let offset_local_input = offset_local + range_start;
 
-        offset_value += offset_local * value.stride(i);
-        offset_input += offset_local_input * input.stride(i);
+        offset_value += offset_local * value.stride(dim);
+        offset_input += offset_local_input * input.stride(dim);
+        offset_remainder = rem;
+
+        comptime![i += 1;]
     }
 
-    input[offset_input] = value[offset_value];
+    input[offset_input / line_size] = value[offset_value / line_size];
 }
 
 pub(crate) fn slice_assign<R: CubeRuntime, E: CubeElement>(
     tensor: CubeTensor<R>,
     indices: &[Range<usize>],
     value: CubeTensor<R>,
 ) -> CubeTensor<R> {
+    let client = tensor.client.clone();
     let tensor = match tensor.can_mut() {
         true => tensor,
         false => tensor.copy(),
     };
     let ndims = tensor.shape.num_dims();
-    let mut indices_sequence = SequenceArg::<R, u32>::new();
+
+    let line_size = if tensor.strides[ndims - 1] == 1 {
+        let last = indices.get(ndims - 1).cloned().unwrap_or(Range {
+            start: 0,
+            end: tensor.shape.dims[ndims - 1],
+        });
+        let shape = last.end - last.start;
+        let offset = last.start;
+        *R::supported_line_sizes()
+            .iter()
+            .filter(|it| {
+                let it = **it as usize;
+                shape % it == 0 && strides_compatible(&tensor.strides, it) && offset % it == 0
+            })
+            .max()
+            .unwrap_or(&1)
```

Comment:
Could we use a built-in function for that? like `tensor_line_size_parallel` or `tensor_line_size_perpendicular`

https://docs.rs/cubecl/latest/cubecl/fn.tensor_line_size_perpendicular.html
https://docs.rs/cubecl/latest/cubecl/fn.tensor_line_size_parallel.html

Reply:
I don't think so because it needs to be compatible with the shape like usual, but also additionally compatible with the offset and slice length.